test/TensorFlowRuntime/tensor_autodiff_runtime.swift

```diff
@@ -121,6 +121,20 @@ TensorADTests.testAllBackends("mean") {
   expectEqual(expected, meanGradAlongAxes(input))
 }
 
+TensorADTests.testAllBackends("variance") {
+  let varianceGradScalar = gradient { (a: Tensor<Float>) in a.variance() }
+  // let varianceGradSqueezingAxes = gradient { (a: Tensor<Float>) in a.variance(squeezingAxes: 0, 1) }
+  let varianceGradAlongAxes = gradient { (a: Tensor<Float>) in a.variance(alongAxes: 0, 1) }
+
+  // TODO: Fix variance derivative calculation.
+  // PyTorch has: `expected == [[-0.75, -0.25], [ 0.25,  0.75]]`.
+  let input: Tensor<Float> = [[1, 2], [3, 4]]
+  let expected = Tensor<Float>(repeating: -0.75, shape: [2, 2])
```

Comment:
Our `variance` ops are implemented using `mean` and `squared` - no custom derivatives for `variance`.

However, the derivatives don't seem to match Python:
```python
import torch
import torch.nn.functional as F

x = torch.tensor([[1, 2], [3, 4]], dtype=torch.float)
x.requires_grad = True
y = x.var(unbiased=False)
y.backward()

print(x)
# tensor([[1., 2.],
#         [3., 4.]], requires_grad=True)

print(y)
# tensor(1.2500, grad_fn=<VarBackward0>)

print(x.grad)
# tensor([[-0.7500, -0.2500],
#         [ 0.2500,  0.7500]])
# We are not consistent with this.
```

Any ideas on fixing on this? Is a custom `variance` derivative necessary?
(I would find it strange if a custom derivative is necessary for correctness.)

Seeking help: cc @jekbradbury @jph00 @sgugger

Reply:
PyTorch has a custom [`var_backward` implementation](https://github.com/pytorch/pytorch/blob/ab4133397c28d20e9d71b38213cc0b14e68ecf08/tools/autograd/templates/Functions.cpp#L665):
```cpp
Tensor var_backward(const Tensor & grad, const Tensor & self, bool unbiased) {
  return (2.0 / (self.numel() - unbiased)) * grad * (self - self.mean());
}

Tensor var_backward(Tensor grad, const Tensor & self, IntArrayRef dim, bool unbiased, bool keepdim) {
  if (self.dim() == 0) {
    return var_backward(grad, self, unbiased);
  }
  if (!keepdim && self.dim() > 1) {
    grad = unsqueeze_multiple(grad, dim, self.sizes().size());
  }
  return (2.0 / (_safe_size(self.sizes(), dim) - unbiased)) * grad * (self - self.mean(dim, true));
}
```

I'll try to port it to Swift.

I'm still concerned about derivative correctness - a custom derivative shouldn't be necessary for correctness, especially because `variance` isn't super complicated (it just calls `mean` and `squared`).

Reply:
Sorry it's not something that @sgugger or I have looked at before so we don't have any particular insights here.

Reply:
Why don't we hand-calculate and verify a simple case?

Reply:
PyTorch is correct and doing
```
let xGrad = gradient(at: x) {x in return (x - x.mean()).squared().mean()}
```
or
```
let xGrad = gradient(at: x) {x in return {x in return x.variance(alongAxes: 0,1)}}
```
gives me the same result as PyTorch.

Reply:
The PyTorch gradient is correct. This Swift version also gives the right results:
```
@differentiable
func variance(_ x: Tensor<Float>) -> Tensor<Float> {
    let mean = x.mean()
    let squaredDiff = (x - mean).squared()
    return squaredDiff.mean()
}
```